File: javascript/sliding-window/max-sum-subarray-of-size-k.js
const max_sub_array_of_size_k_ta = function(k, arr) {
    result = 0;
    
    for (let i = 0; i < arr.length - k + 1; i++) {
      windowMax = 0;
      for (let j = 0; j < k; j++) {
        windowMax += arr[i+j];
      }
      result = Math.max(result, windowMax);
    } 
    return result;
};

console.log(`Maximum sum of a subarray of size K: ${max_sub_array_of_size_k_ta(3, [2, 1, 5, 1, 3, 2])}`);
console.log(`Maximum sum of a subarray of size K: ${max_sub_array_of_size_k_ta(2, [2, 3, 4, 1, 5])}`);

function max_sub_array_of_size_k(k, arr) {
    let maxSum = 0,
      windowSum = 0;
  
    for (i = 0; i < arr.length - k + 1; i++) {
      windowSum = 0;
      for (j = i; j < i + k; j++) {
        windowSum += arr[j];
      }
      maxSum = Math.max(maxSum, windowSum);
    }
    return maxSum;
  }
  
  
console.log(`Maximum sum of a subarray of size K: ${max_sub_array_of_size_k(3, [2, 1, 5, 1, 3, 2])}`);
console.log(`Maximum sum of a subarray of size K: ${max_sub_array_of_size_k(2, [2, 3, 4, 1, 5])}`);
  
function max_sub_array_of_size_k_improved(k, arr) {
    let maxSum = 0,
      windowSum = 0,
      windowStart = 0;
  
    for (window_end = 0; window_end < arr.length; window_end++) {
      windowSum += arr[window_end]; // add the next element
      // slide the window, we don't need to slide if we've not hit the required window size of 'k'
      if (window_end >= k - 1) {
        maxSum = Math.max(maxSum, windowSum);
        windowSum -= arr[windowStart]; // subtract the element going out
        windowStart += 1; // slide the window ahead
      }
    }
    return maxSum;
  }
  
  
  console.log(`Maximum sum of a subarray of size K: ${max_sub_array_of_size_k_improved(3, [2, 1, 5, 1, 3, 2])}`);
  console.log(`Maximum sum of a subarray of size K: ${max_sub_array_of_size_k_improved(2, [2, 3, 4, 1, 5])}`);

File: python/sliding-window/max-sum-subarray-of-size-k.py
# Maximum Sum Subarray of Size K (easy)

# Problem Statement#

# Given an array of positive numbers and a positive number ‘k,’ find the maximum sum of any contiguous subarray of size ‘k’.

# Example 1:

# Input: [2, 1, 5, 1, 3, 2], k=3 
# Output: 9
# Explanation: Subarray with maximum sum is [5, 1, 3].

# Example 2:

# Input: [2, 3, 4, 1, 5], k=2 
# Output: 7
# Explanation: Subarray with maximum sum is [3, 4].

# This is me just playing around and re-learning python syntax with a pattern I know
def max_sub_array_of_size_k_ta(k, arr):
    max_sum = 0
    window_sum = 0
    window_start = 0
    
    for window_end in range(len(arr)):
        window_sum += arr[window_end]
        # window size limit
        if window_end >= k-1:
            max_sum = max(max_sum, window_sum)
            # shrink the window
            window_sum -= arr[window_start]
            window_start += 1
            
    return max_sum

def max_sub_array_of_size_k(k, arr):
    max_sum = 0
    window_sum = 0

    for i in range(len(arr) - k + 1):
        window_sum = 0
        for j in range(i, i+k):
            window_sum += arr[j]
        max_sum = max(max_sum, window_sum)
    return max_sum

def max_sub_array_of_size_k_optimal(k, arr):
    max_sum , window_sum = 0, 0
    window_start = 0

    for window_end in range(len(arr)):
        window_sum += arr[window_end]  # add the next element
        # slide the window, we don't need to slide if we've not hit the required window size of 'k'
        if window_end >= k-1:
            max_sum = max(max_sum, window_sum)
            window_sum -= arr[window_start]  # subtract the element going out
            window_start += 1  # slide the window ahead
    return max_sum

def main():
    print("Maximum sum of a subarray of size K: " + str(max_sub_array_of_size_k_ta(3, [2, 1, 5, 1, 3, 2])))
    print("Maximum sum of a subarray of size K: " + str(max_sub_array_of_size_k_ta(2, [2, 3, 4, 1, 5])))
    
    print("Maximum sum of a subarray of size K: " + str(max_sub_array_of_size_k(3, [2, 1, 5, 1, 3, 2])))
    print("Maximum sum of a subarray of size K: " + str(max_sub_array_of_size_k(2, [2, 3, 4, 1, 5])))
    
    print("Maximum sum of a subarray of size K: " + str(max_sub_array_of_size_k_optimal(3, [2, 1, 5, 1, 3, 2])))
    print("Maximum sum of a subarray of size K: " + str(max_sub_array_of_size_k_optimal(2, [2, 3, 4, 1, 5])))

main()
